### PR TITLE
Add mixed-precision (FP32) LM output logits

### PR DESCRIPTION
**Pull request by mkhona-nvidia**
- [x] I, the PR author, have personally reviewed every line of this PR.

# What does this PR do?
Allow the language model to compute `output_layer` with logits dtype leaving the accumulator in FP32, rather than casting down to BF16 as default. We use TE's `general_gemm` API to do this and modify autograd to cast the gradients back down to BF16, following Megatron-LM's contract with TE for GEMMs. 

The legacy path performs FP32 accumulation internally but rounds every vocabulary logit to BF16 before cross entropy. Casting those rounded values back to FP32 cannot recover the discarded bits. Requesting an FP32 GEMM output preserves the accumulator result and materially reduces error at every tested reduction width.

The explicit backward cast does not introduce a new gradient contract: it reproduces the cast that autograd already performs through the legacy post-GEMM `logits.float()` operation.

## Tests

- **FP32 output and backward contract:** verifies BF16 input/weight GEMMs produce FP32 logits while input and weight gradients remain BF16 and match the legacy path.
- **Integer bitwise exactness:** uses integer-valued BF16 operands in `[-8, 8]` with `K=512`. Since `K * 8^2 = 32,768 < 2^24`, every product and partial sum is exactly representable in FP32; the test compares raw FP32 words against a PyTorch FP32 reference with no tolerance.
- **Plain Transformer Engine parity:** compares the complete Megatron wrapped linear path against a direct `transformer_engine.pytorch.cpp_extensions.general_gemm` call configured with the same FP32 output contract, requiring bitwise-identical outputs.
- **Backward regression sweep:** applies the same FP32 upstream gradient to legacy and mixed-output paths across hidden sizes 512–16,384; BF16 input and weight gradients are bitwise identical at every width (Slurm job `2825327`).
- **Numerical accuracy sweep:** across hidden sizes 512–16,384, FP32-output logit RMSE improves by 181.2×–3,323.6× and cross-entropy absolute error is lower at every tested width relative to legacy BF16 logits (Slurm job `2825327`).
- **Kernel benchmark:** compares preallocated BF16-output `torch.matmul` against Transformer Engine `general_gemm` with an FP32 output over 15 shapes (`M` in 16, 128, 1,024; `K` in 1,024–16,384; `N=32,768`). The median overhead was 5.4% overall, 2.2% for `K >= 4,096`, and 1.9% for `K >= 8,192`; 11/15 shapes were within 10% and the full range was 0.3%–15.9% (NVIDIA GB300, Slurm job `2825411`).
- **CLI contract:** verifies `--output-logit-dtype` defaults to `bf16`, accepts only `bf16` and `fp32`, and rejects `fp16`.

### GEMM kernel benchmark

| Tokens (M) | Hidden (K) | BF16 output | FP32 output | Overhead |
|---:|---:|---:|---:|---:|
| 16 | 1,024 | 0.022 ms | 0.025 ms | +15.8% |
| 16 | 2,048 | 0.025 ms | 0.026 ms | +5.4% |
| 16 | 4,096 | 0.046 ms | 0.046 ms | +1.5% |
| 16 | 8,192 | 0.086 ms | 0.087 ms | +1.0% |
| 16 | 16,384 | 0.193 ms | 0.194 ms | +0.3% |
| 128 | 1,024 | 0.022 ms | 0.026 ms | +15.9% |
| 128 | 2,048 | 0.030 ms | 0.033 ms | +10.3% |
| 128 | 4,096 | 0.053 ms | 0.057 ms | +6.6% |
| 128 | 8,192 | 0.104 ms | 0.107 ms | +2.9% |
| 128 | 16,384 | 0.192 ms | 0.195 ms | +1.6% |
| 1,024 | 1,024 | 0.054 ms | 0.062 ms | +14.9% |
| 1,024 | 2,048 | 0.087 ms | 0.093 ms | +7.2% |
| 1,024 | 4,096 | 0.158 ms | 0.168 ms | +6.2% |
| 1,024 | 8,192 | 0.309 ms | 0.317 ms | +2.5% |
| 1,024 | 16,384 | 0.659 ms | 0.674 ms | +2.2% |

The benchmark uses preallocated output buffers, 10 warmup pairs, 30 alternating-order CUDA-event samples per kernel, and reports medians. The vocabulary dimension is 32,768 for every shape. The larger-width results show that preserving FP32 GEMM output has low kernel-level overhead for representative LM heads.

| Hidden size | Legacy logit RMSE | FP32-output RMSE | RMSE improvement | Legacy CE abs. error | FP32-output CE abs. error | Backward grads |
|---:|---:|---:|---:|---:|---:|:---:|
| 512 | 1.690e-03 | 5.084e-07 | 3323.6× | 2.804e-04 | 0.000e+00 | Exact |
| 1,024 | 1.674e-03 | 1.028e-06 | 1628.6× | 4.301e-04 | 9.537e-07 | Exact |
| 2,048 | 1.674e-03 | 2.156e-06 | 776.5× | 2.527e-04 | 1.907e-06 | Exact |
| 4,096 | 1.664e-03 | 4.464e-06 | 372.9× | 4.387e-05 | 2.861e-06 | Exact |
| 8,192 | 1.660e-03 | 9.161e-06 | 181.2× | 8.106e-05 | 5.722e-06 | Exact |
| 16,384 | 1.663e-03 | 2.862e-06 | 581.0× | 4.501e-04 | 9.537e-07 | Exact |

### Integer bitwise exactness results

| Hidden size | FP32-output bit mismatches | Legacy value mismatches | Result |
|---:|---:|---:|:---:|
| 512 | 0 / 16,384 | 6,767 / 16,384 | Exact |
| 1,024 | 0 / 16,384 | 8,557 / 16,384 | Exact |
| 2,048 | 0 / 16,384 | 10,027 / 16,384 | Exact |
| 4,096 | 0 / 16,384 | 11,335 / 16,384 | Exact |
| 8,192 | 0 / 16,384 | 12,443 / 16,384 | Exact |
| 16,384 | 0 / 16,384 | 13,311 / 16,384 | Exact |

Both operands are BF16 tensors containing integers in `[-8, 8]`. For every width, `K * 8^2 < 2^24`, so the products and possible partial sums are exactly representable in FP32. The raw FP32 output words matched the FP32 PyTorch reference with no tolerance at every width (NVIDIA GB300).

## Method

- Inputs and output-layer weights are BF16; the reference uses their exactly represented FP32 values.
- Legacy path: PyTorch BF16 output-layer GEMM, followed by the existing FP32 logits cast.
- New path: Transformer Engine `general_gemm` with `out_dtype=torch.float32`.
- Reference: PyTorch FP32 linear projection with TF32 disabled.
- Shape sweep: `16 x K` activations, `32,768 x K` vocabulary weights, K in `512`, `1024`, `2048`, `4096`, `8192`, `16384`.
- Hardware: NVIDIA GB300; seed `1234`; commit `27ef2b465. 

## Issue tracking

For PRs from open-source community contributors:

- **New features**: a linked issue is **required**. Please open a [feature request](https://github.com/NVIDIA/Megatron-LM/issues/new?template=feature_request.md) and reference it here before submitting the PR.
- **Small updates (bug fixes, minor improvements)**: a linked issue is **recommended** and will accelerate the PR review process.

Linked issue: 

### Pre-checks

- [ ] I have added relevant unit tests
- [ ] I have added relevant functional tests
- [ ] I have added proper typing to my code [Typing guidelines](https://docs.python.org/3/library/typing.html)
- [ ] I have added relevant documentation
- [ ] I have run the [autoformatter.sh](https://github.com/NVIDIA/Megatron-LM/blob/main/tools/autoformat.sh) on my PR

### Code review

Feel free to message or comment @NVIDIA/mcore-oncall to help accelerate your merge into main. The less complex your PR is, the faster it will be approved and merged!

All PRs start as **draft**. If you open a non-draft PR, it will be automatically converted to draft.

#### Step 1: Mark PR as "Ready for Review"

1. When your PR is ready, click **Ready for Review**.
2. An oncall reviewer is auto-assigned and expert reviewers are notified based on your changes.
   - Some PRs may jump straight to step 2. This is determined by `.github/CODEOWNERS`.

:warning: Only mark as ready once merge-conflicts are resolved and the CI is passing.
Final Review might get declined if these requirements are not fulfilled.

#### Step 2: Final Review

For PRs that change `megatron/core`, once all expert reviewers have approved, the `Final Review` label is applied **automatically** and final reviewers are assigned.

For PRs outside `megatron/core`, this step is skipped.

#### Step 3: Approved

Once all required reviewers have approved, the `Approved` label is applied **automatically**.

### Merge

Any member of [mcore-engineers](https://github.com/orgs/NVIDIA/teams/mcore-engineers) will be able to merge your PR.

### CODE REVIEW

**Review by deepakn94 — APPROVED** [review]

**Review by AAnoosheh — APPROVED** [review]

**Review by Phlip79 — APPROVED** [review]

**Review by cuichenx — APPROVED** [review]

**Comment by deepakn94** [comment]
/ok to test [2800628](https://github.com/NVIDIA/Megatron-LM/pull/6252/commits/2800628eaaf0e0e3e0ece0be59d8a66853791c9b)

**Comment by deepakn94** [comment]
/ok to test [6e10e6b](https://github.com/NVIDIA/Megatron-LM/pull/6252/commits/6e10e6b6f281513812d8d90b90975dcc08e12056)

**Review by tdene — APPROVED** [review]

**Review by maanug-nv — APPROVED** [review]

File: megatron/core/inference/model_inference_wrappers/abstract_model_inference_wrapper.py [hunk 1]
```diff
@@ -220,7 +224,7 @@ def forward_pass_with_pipeline_parallel(
             logits = output_tensor
 
             # Explicitly cast logits to expected dtype
-            logits = logits.to(self.config.params_dtype)
+            logits = logits.to(self.logit_dtype)
```
Comment (sidsingh-nvidia) [inline megatron/core/inference/model_inference_wrappers/abstract_model_inference_wrapper.py]: Would this be a no-op with fp32 logit dtype?

**Review by sidsingh-nvidia — APPROVED** [review]

File: megatron/core/inference/model_inference_wrappers/abstract_model_inference_wrapper.py (hunk 1 above)
Comment (mkhona-nvidia) [reply megatron/core/inference/model_inference_wrappers/abstract_model_inference_wrapper.py]: @santhnm2

**Comment by deepakn94** [comment]
/ok to test [c2ca628](https://github.com/NVIDIA/Megatron-LM/pull/6252/commits/c2ca628caa0b3811720c930cb1a8d58abc96fdad)
